File: 42583.cpp
// 다리를 지나는 트럭

#include <string>
#include <vector>
#include <queue>

using namespace std;

int solution(int bridge_length, int weight, vector<int> truck_weights) {
    queue<int> q;
    int answer = 0;
    int sum = 0;
    int truckIdx = 0;

    while (1) {
        answer++;

        if (q.size() == bridge_length) {
            sum -= q.front();
            q.pop();
        }

        if (sum + truck_weights[truckIdx] <= weight) {
            if (truckIdx == truck_weights.size() - 1) {
                answer += bridge_length;
                break;
            }

            q.push(truck_weights[truckIdx]);
            sum += truck_weights[truckIdx];
            truckIdx++;
        }
        else {
            q.push(0);
        }
    }
    return answer;
}

File: 42583.py
# 다리를 지나는 트럭
# 2021-09-21

def solution(bridge_length, weight, truck_weights):
    answer, stack = 0, [0] * bridge_length
    while len(stack):
        stack.pop(0)
        if truck_weights:
            if sum(stack) + truck_weights[0] <= weight:
                stack.append(truck_weights.pop(0))
            else:
                stack.append(0)
        else:
            answer += bridge_length
            break
        answer += 1

    return answer

print(solution(2, 10, [7,4,5,6]))